static/app/components/events/eventTags/eventTagsTree.tsx

```diff
@@ -0,0 +1,202 @@
+import styled from '@emotion/styled';
+
+import EventTagsContent from 'sentry/components/events/eventTags/eventTagContent';
+import {space} from 'sentry/styles/space';
+import type {EventTag} from 'sentry/types';
+import {generateQueryWithTag} from 'sentry/utils';
+import {useLocation} from 'sentry/utils/useLocation';
+import useOrganization from 'sentry/utils/useOrganization';
+
+const MAX_TREE_DEPTH = 4;
+interface TagTree {
+  [key: string]: TagTreeContent;
+}
+interface TagTreeContent {
+  subtree: TagTree;
+  value: string;
+  // These will be omitted on pseudo tags (see addToTagTree)
+  meta?: Record<any, any>;
+  originalTag?: EventTag;
+}
+
+function addToTagTree(
+  tree: TagTree,
+  tag: EventTag,
+  meta: Record<any, any>,
+  originalTag: EventTag
+): TagTree {
+  const branchMatches = tag.key.match(/\./g) ?? [];
+
+  const hasInvalidBranchCount =
+    branchMatches.length <= 0 || branchMatches.length > MAX_TREE_DEPTH;
+  const hasInvalidBranchSequence = (tag.key.match(/\.{2,}/g) ?? []).length > 0;
+
+  // Ignore tags with 0, or >4 branches, as well as sequential dots (e.g. 'some..tag')
+  if (hasInvalidBranchCount || hasInvalidBranchSequence) {
+    tree[tag.key] = {value: tag.value, subtree: {}, meta, originalTag};
+    return tree;
+  }
+  // E.g. 'device.model.version'
+  const splitIndex = tag.key.indexOf('.'); // 6
+  const trunk = tag.key.slice(0, splitIndex); // 'device'
+  const branch = tag.key.slice(splitIndex + 1); // 'model.version'
+
+  if (tree[trunk] === undefined) {
+    // We need to define the default as an empty space for alignment
+    tree[trunk] = {value: ' ', subtree: {}};
+  }
+  // Recurse with a pseudo tag, e.g. 'model', to create nesting structure
+  const pseudoTag = {
+    key: branch,
+    value: tag.value,
+  };
+  tree[trunk].subtree = addToTagTree(tree[trunk].subtree, pseudoTag, meta, originalTag);
+  return tree;
+}
+
+interface TagsTreeKeysProps {
+  content: TagTreeContent;
+  tag: string;
+}
+
+function TagTreeKeys({content, tag}: TagsTreeKeysProps) {
```

Comment:
This and `TagTreeValues` are both recursive components. They needed their own components rather than just being part of the render of `<EventTagTree>` since they needed to follow the subgrid generated for each row. This also has the added benefit of one component to hold the entire subtree, preventing a situation where a parent tag is in a different column than its child (if we end up going with that UI)